src/decoder/cuda-decoder-utils.h

```diff
@@ -0,0 +1,161 @@
+// decoder/cuda-decoder-utils.h
+
+// Copyright      2018  Zhehuai Chen
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef KALDI_CUDA_DECODER_UTILS_H_
+#define KALDI_CUDA_DECODER_UTILS_H_
+
+
+#include <cuda_runtime_api.h>
+#include "cuda_runtime.h"
+#include "util/stl-utils.h"
+#include "omp.h"
+#include <algorithm>
+#include <float.h>
+#include <math.h>
+#include <cooperative_groups.h>
+#include "math_constants.h"
+#include "omp.h"
+
+// cuda macro
+
+#ifdef __CUDACC__
+#define HOST __host__
```

Comment:
This looks a little strange to me. "nvcc" can compile host code just fine. I suspect you can accomplish what you want simply by prefixing your function declarations with "__host__ __device__"

Reply:
it's because we do different behaviours in host and device

Reply:
If you need different behaviors between host and device, I suggest using something like this:

```
#ifdef __CUDA_ARCH__
// GPU code
#else
// CPU code
#endif
```

__device__ and __host__ keywords are orthogonal to whether you are using different behaviors. Since you are using nvcc to compile this code (I assume), you can simply use those keywords directly.

Reply:
We might be able to arrange to just not compile this code at all if nvcc is not available.  After all, the whole thing doesn't make sense in that case.  If these ifdefs came from Justin, though, let me know- he may have some reason I'm not aware of for doing it.

Reply:
Oh, I see--it looks like nvcc may compile the same file twice, once to generate code to run on the device, and once to run it on the CPU.

Reply:
I have similar understanding here.